codepropertygraph/src/test/scala/io/shiftleft/codepropertygraph/cpgloading/DiffGraphTest.scala

```diff
@@ -0,0 +1,105 @@
+package io.shiftleft.codepropertygraph.cpgloading
+
+import gremlin.scala._
+import io.shiftleft.OverflowDbTestInstance
+import io.shiftleft.codepropertygraph.Cpg
+import io.shiftleft.codepropertygraph.generated._
+import io.shiftleft.codepropertygraph.generated.nodes.{NewNode, StoredNode}
+import io.shiftleft.passes.DiffGraph
+import org.scalatest.{Matchers, WordSpec}
+
+class DiffGraphTest extends WordSpec with Matchers {
+  "should be able to build an inverse DiffGraph" in {
+    withTestOdb { g =>
+      // setup existing graph
+      // add x and y nodes to graph
+      val x = g.addVertex(NodeTypes.UNKNOWN)
+      val y = g.addVertex(NodeTypes.UNKNOWN)
+      x.property(NodeKeyNames.CODE, "x")
+      y.property(NodeKeyNames.CODE, "old y code")
+      val x2y = x.addEdge(EdgeTypes.CONTAINS_NODE, y, EdgeKeyNames.LOCAL_NAME, "old edge attr")
+      // make diffgraph
+      var diffBuilder = DiffGraph.newBuilder
+      // add a b c nodes to the DiffGraph.Builder
+      val a = makeNode(g, "a")
+      diffBuilder.addNode(a)
+      val b = makeNode(g, "b")
+      diffBuilder.addNode(b)
+      val c = makeNode(g, "c")
+      diffBuilder.addNode(c)
+      // add edge a -> b -> c to the DiffGraph.Builder
+      makeEdgeBetweenNewNodes(g, diffBuilder, a, b)
+      makeEdgeBetweenNewNodes(g, diffBuilder, b, c)
+      // add edge from existing node "x" to new node "a" to the builder
+      diffBuilder.addEdgeFromOriginal(x.asInstanceOf[StoredNode], a, EdgeTypes.AST)
+      // modify property of existing node "y"
+      diffBuilder.addNodeProperty(y.asInstanceOf[StoredNode], NodeKeyNames.ORDER, new Integer(123))
+      diffBuilder.addNodeProperty(y.asInstanceOf[StoredNode], NodeKeyNames.CODE, "new y code")
+
+      diffBuilder.addEdgeProperty(x2y, EdgeKeyNames.LOCAL_NAME, "new edge attr")
+      val diffGraph = diffBuilder.build()
+      // apply diffgraph with undoable = true
+      val appliedDiffGraph = DiffGraph.Applier.applyDiff(diffGraph, g.graph, true)
+      val inverseDiffGraph = appliedDiffGraph.inverseDiffGraph.get
+      val changes = inverseDiffGraph.iterator.toList
+      import DiffGraph.Change._
+      val List(
+        SetEdgeProperty(_, EdgeKeyNames.LOCAL_NAME, "old edge attr"), // restore old edge property value
+        SetNodeProperty(_, NodeKeyNames.CODE, "old y code"), // restore old Y property value
+        RemoveNodeProperty(_, NodeKeyNames.ORDER), // remove newly added property
+        RemoveEdge(_), // remove x -> a
+        RemoveEdge(_), // remove b -> c
+        RemoveEdge(_), // remove a -> b
+        RemoveNode(_), // remove c
+        RemoveNode(_), // remove b
+        RemoveNode(_) // remove a
+      ) = changes
```

Comment:
does this test assert anything?

Reply:
@mpollmeier Yes, the pattern matching will throw an exception if the structure of `changes` var does not correspond to the pattern above. This could be done with asserts and instanceOf checks of each list element but I think this is more readable

Reply:
ah right, ok!